partiql-tests-data/eval/primitives/coll-aggregate-function.ion

```diff
@@ -344,6 +344,366 @@ coll_min::[
   },
 ]
 
+coll_any::[
+  {
+    name:"COLL_ANY bag literals",
+    statement:"COLL_ANY(<< false, true, false >>)",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:true
+    }
+  },
+  {
+    name:"COLL_ANY list expressions",
+    statement:"COLL_ANY([ 1 < 5, false, 5 IS NULL ])",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:true
+    }
+  },
+  {
+    name:"COLL_ANY single true",
+    statement:"COLL_ANY(<<true>>)",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:true
+    }
+  },
+  {
+    name:"COLL_ANY single false",
+    statement:"COLL_ANY([false])",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:false
+    }
+  },
+  {
+    name:"COLL_ANY nulls with true",
+    statement:"COLL_ANY(<< NULL, 2<3, MISSING, false >>)",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:true
+    }
+  },
+  {
+    name:"COLL_ANY nulls with false",
+    statement:"COLL_ANY([ 2>3, NULL ])",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:false
+    }
+  },
+  {
+    name:"COLL_ANY nulls only",
+    statement:"COLL_ANY([NULL, MISSING]])",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:null
+    }
+  },
+  {
+    name:"COLL_ANY some empty",
+    statement:"COLL_ANY(<< >>)",
+    assert:{
+      evalMode:[EvalModeCoerce, EvalModeError],
+      result:EvaluationSuccess,
+      output:null
+    }
+  },
+  {
+    name:"COLL_ANY one non-bool, non-unknown",
+    statement:"COLL_ANY(<< true, 5, true >>)",
+    assert:[
+      {
+        evalMode: EvalModeError,
+        result:EvaluationFail,
+      },
+      {
+        evalMode:EvalModeCoerce,
+        result:EvaluationSuccess,
+        output:$missing::null
+      }
+    ]
+  },
+  {
+    name:"COLL_ANY all non-bool, non-unknown",
+    statement:"COLL_ANY([ 5, 'hello', 3.14])",
+    assert:[
+      {
+        evalMode: EvalModeError,
+        result:EvaluationFail,
+      },
+      {
+        evalMode:EvalModeCoerce,
+        result:EvaluationSuccess,
+        output:$missing::null
```

Comment:
You bring up some good points. I had a few reasons I was outputting `MISSING` rather than `NULL` for these `COLL_` functions.

1. I was defining the `COLL_` functions to follow the general behavior for builtin functions. That is, if there is a mistyped argument, we should return missing for the entire function call in permissive mode (error in strict mode). See [section 7.1](https://partiql.org/assets/PartiQL-Specification.pdf#subsection.7.1) of the spec:

> All functions return `MISSING` when they input data whose types do not conform to the input argument types. 

That's why I believed `MISSING` would be output in this case for the function expression since the input list contained non-boolean, non-null/missing values.

2. Returning `MISSING` in permissive mode can be helpful for the user to know that a data type mismatch had occurred rather than ignoring the mistyped argument.

3. For your point about boolean expressions (e.g. `AND`, `OR`, and also `IN`) following 3-value logic, I defined the `COLL_` following the builtin function semantics rather than the boolean expression semantics. I felt it would be more consistent if all the `COLL_` followed the same data type mismatch behavior.

E.g. `COLL_SUM(['non-integer'] -> MISSING` and `COLL_ANY(['non-boolean', true]) -> MISSING` both returning missing on a mistyped element.

4. Returning `MISSING` for `COLL_ANY`/`COLL_SUM`/`COLL_EVERY` is consistent with the rewrite of SQL aggregates mentioned in [section 11.2.2](https://partiql.org/assets/PartiQL-Specification.pdf#subsubsection.11.2.2).

Consider a PartiQL query as follows:

```
SELECT a, ANY(a) AS agg_result
FROM <<{'a': true}, {'a': false}, {'a': 5}>> AS t
GROUP BY a GROUP AS g
```

The expected result in permissive mode would be:

```
<<
  {
    'a': false,
    'agg_result': true
  },
  {
    'a': true,
    'agg_result': true
  },
  {
    'a': 5,
    -- 'agg_result' omitted since `MISSING` attributes are ignored
  }
>>
```

If we rewrite to use `COLL_ANY` rather than `ANY`, the rewritten query would be something like:
```
SELECT a, COLL_ANY(SELECT VALUE p.t.a FROM g AS p) AS agg_result
FROM <<{'a': true}, {'a': false}, {'a': 5}>> AS t
GROUP BY a GROUP AS g
```

The result of the subquery for the `a = 5` group would be `[5]`. So the result of `COLL_ANY([5])` would have to be `MISSING` in this case.

---

I guess to summarize, I made the following assumptions to determine `MISSING` was returned rather than `NULL`:
- `COLL_` functions all act like other builtin call functions -- return `MISSING` on data type mismatches
- `COLL_ANY`/`COLL_SOME`/`COLL_EVERY` don't follow the boolean expression 3-value logic but are consistent with builtin functions
- Assuming SQL aggregates return `MISSING` on data type mismatches (which is one underspecified aspect part of this spec issue https://github.com/partiql/partiql-spec/issues/50)